core/matrix/csr.cpp

```diff
@@ -618,6 +621,83 @@ void Csr<ValueType, IndexType>::write(mat_data& data) const
 }
 
 
+template <typename ValueType, typename IndexType, typename TransformClosure>
+std::pair<std::unique_ptr<Csr<ValueType, IndexType>>,
+          typename Csr<ValueType, IndexType>::permuting_reuse_info>
+transform_reusable(const Csr<ValueType, IndexType>* input, gko::dim<2> out_size,
+                   size_type nnz, TransformClosure closure)
+{
+    using FloatIndexType =
+        std::conditional_t<std::is_same_v<IndexType, int32>, float, double>;
+    static_assert(sizeof(FloatIndexType) == sizeof(IndexType));
+    static_assert(alignof(FloatIndexType) == alignof(IndexType));
+    auto exec = input->get_executor();
+    auto in_size = input->get_size();
+    auto transformed = Csr<ValueType, IndexType>::create(exec, out_size, nnz);
+    // transform matrix with integer values from 0 to nnz - 1 reinterpret_cast
+    // as float
+    array<IndexType> iota_values{exec, nnz};
+    exec->run(csr::make_fill_seq_array(iota_values.get_data(), nnz));
+    auto iota_float_view = make_array_view(
+        exec, nnz, reinterpret_cast<FloatIndexType*>(iota_values.get_data()));
```

Comment:
TBH I'm a bit skeptical about this. Isn't this UB?

Reply:
Yes, but we are provoking the same kind of nasal demons elsewhere - we don't really care about object provenance for index and value types, because we allocate the data using a pointer allocated via an untyped `::operator new`. The point here is that the compiler can't do anything with this information, because we always pass it through the ABI boundary between the ginkgo and ginkgo_... libraries, only the kernels actually access these values through regular typed pointers instead of a memcpy.

Reply:
Thanks for the clarification. If the bit patterns match, then I guess we can live with it.